Dockerfile.gcc

```diff
@@ -96,17 +96,14 @@ ENV SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH}"
 # ensure CONFIG_BPF_KPROBE_OVERRIDE is not set
 RUN ! grep -q -E 'CONFIG_BPF_KPROBE_OVERRIDE=[y|m]' .config || (echo "Cannot set this option for security reasons"; false)
 
+RUN make -j$(nproc) mrproper
+RUN make O=/kernel-out ${KERNEL_CONFIG}
 RUN --mount=type=cache,target=/root/.cache/ccache,id=kernel-ccache-${TARGETARCH} \
-    ccache -z && \
-    echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} and CROSS_COMPILE=${CROSS_COMPILE}" && \
-    make -j$(nproc) mrproper \
-    && make O=/kernel-out ${KERNEL_CONFIG} \
-    && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) prepare \
-    && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) \
-    && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) modules \
+    echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} and CROSS_COMPILE=${CROSS_COMPILE}" \
+    && make CC="${CC}" O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) prepare bzImage modules \
     && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) modules_install INSTALL_MOD_STRIP=1 \
         INSTALL_MOD_PATH=/tmp/kernel-modules \
-    && ccache -s
+    && ccache -s | tee -a /ccache-stats.txt
```

Comment:
Why ccache is not enabled? The CC is defined here: https://github.com/lf-edge/eve-kernel/blob/eve-kernel-amd64-v6.1.38-generic/Dockerfile.gcc#L29 with the ccache command included.... at least for arm (cross-compiling) I can ensure it's working...

Reply:
I didn't check the process, but by applying the following:

```diff
diff --git a/Dockerfile.gcc b/Dockerfile.gcc
index 45286bc2ac9b..9aa68ff823a5 100644
--- a/Dockerfile.gcc
+++ b/Dockerfile.gcc
@@ -98,7 +98,7 @@ RUN ! grep -q -E 'CONFIG_BPF_KPROBE_OVERRIDE=[y|m]' .config || (echo "Cannot set
 
 RUN --mount=type=cache,target=/root/.cache/ccache,id=kernel-ccache-${TARGETARCH} \
     ccache -z && \
-    echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} and CROSS_COMPILE=${CROSS_COMPILE}" && \
+    echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} , CC=${CC} and CROSS_COMPILE=${CROSS_COMPILE}" && \
     make -j$(nproc) mrproper \
     && make O=/kernel-out ${KERNEL_CONFIG} \
     && make O=/kernel-out LOCALVERSION="-${LOCALVERSION}" -j$(nproc) prepare \
```

I've got:

```
#17 0.164 Statistics zeroed
#17 0.164 Building kernel for amd64 with ARCH=x86_64 , CC=ccache gcc and CROSS_COMPILE=
```

so I suppose it should work... will check the process as well...

Reply:
@rene an easy way to test this is do this change:

```
christoph@build-server ~/p/eve-kernel (eve-kernel-amd64-v6.1.68-generic)> git diff
diff --git a/Dockerfile.gcc b/Dockerfile.gcc
index 9e0e780fcdd3..141988cfdcbc 100644
--- a/Dockerfile.gcc
+++ b/Dockerfile.gcc
@@ -96,6 +96,7 @@ ENV SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH}"
 # ensure CONFIG_BPF_KPROBE_OVERRIDE is not set
 RUN ! grep -q -E 'CONFIG_BPF_KPROBE_OVERRIDE=[y|m]' .config || (echo "Cannot set this option for security reasons"; false)
 
+ENV CC="asdf ccache gcc"
 RUN --mount=type=cache,target=/root/.cache/ccache,id=kernel-ccache-${TARGETARCH} \
     ccache -z && \
     echo "Building kernel for ${TARGETARCH} with ARCH=${ARCH} and CROSS_COMPILE=${CROSS_COMPILE}" && \
``` 

on `eve-kernel-amd64-v6.1.68-generic` branch and build it with: `docker build -f Dockerfile.gcc .` and then you see it does not fail while building the kernel.

Reply:
Understood. The problem is not docker, the variable it's exported, but the Makefile from the kernel doesn't consider an environment CC, so as it's solved by this PR, it must be defined at make command line to get priority over the one declared inside the Makefile (which is `$(CROSS_COMPILE)gcc`), and this explains why it was working for arm, because we define the ccache at CROSS_COMPILE and not on CC....

Reply:
@christoph-zededa , please, don't forge to provide the speedup cache solution (break into another Dockerfile) for arm as well....

Reply:
unfortunately breaking it into two Dockerfiles does only work on some special versions of docker it seems (especially the ones that do not use the more modern buildkit)